### DISCUSSION TITLE
Error: Could not resolve @prisma/client despite the installation that we just tried

### DISCUSSION THREAD

**Discussion by 269652** [prisma/prisma#14572] [Q&A] [answered]
### Bug description

Running `npx prisma generate` yields the following error. 

```
Error: Could not resolve @prisma/client despite the installation that we just tried. 
Please try to install it by hand with npm install @prisma/client and rerun npx prisma generate
```

Installing the prisma client and running `npx prisma generate` again doesn't change the error message. 
Yesterday everything was working fine. 

### How to reproduce

1. `npm i @prisma/client`
2. `npx prisma generate`
3. See error

### Expected behavior

The prisma client should be generated.

### Prisma information

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model states {
  id    String @id
  state Json?  @db.Json
}

### Environment & setup

- OS: Windows
- Database: PostgreSql
- Node.js version: v14.17.0

### Prisma Version

```
4.1.1
```

**Comment by janpio** [comment]
What changed since yesterday?
What does your Node project setup look like?
Are you running the commands in the same folder as before?
Is `@prisma/client` installed as a depency in that project?

**Comment by 269652** [comment, marked as answer]
I think I that a node_modules folder in the parent folder was causing the issue. It seems i accidentally ran `npm i some-package` in the parent folder. I don't know why this causes Prisma not being able to resolve the path to its installation. 

It was correctly installed under /parent/project/node_modules/.

It seems that `npx prisma generate` doesn't correctly resolve the path if there's a parent folder with node_modules. 

I'll try and setup a reproducible example tomorrow.

**Reply by vladi-stevanovic** [reply]
👋 Hello @C5H8NNaO4 I'm going through pending GitHub Discussions to check if you need any help. Is this issue still open?
Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! 🙌

**Reply by tessro** [reply]
We ran into this, and the root cause turned out to be two different versions of `@prisma/client` in our monorepo due to desync between various `package.json` files.

**Reply by yuranos** [reply]
Having this same issue running through an example project for [typegraphql](https://github.com/prisma/prisma-examples/tree/latest/typescript/graphql-typegraphql).
```
(master) $ npx prisma migrate dev
Prisma schema loaded from prisma/schema.prisma
Datasource "db": SQLite database "dev.db" at "file:./dev.db"

Already in sync, no schema change or pending migration was found.

Running generate... (Use --skip-generate to skip the generators)
➤ YN0000: ┌ Resolution step
➤ YN0002: │ ts-node-dev@npm:2.0.0 [6cac6] doesn't provide @types/node (p09428), requested by ts-node
➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
➤ YN0000: └ Completed
➤ YN0000: ┌ Link step
➤ YN0000: │ ESM support for PnP uses the experimental loader API and is therefore experimental
➤ YN0000: └ Completed
➤ YN0000: Done with warnings in 0s 280ms
➤ YN0000: ┌ Resolution step
➤ YN0002: │ ts-node-dev@npm:2.0.0 [6cac6] doesn't provide @types/node (p09428), requested by ts-node
➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
➤ YN0000: └ Completed
➤ YN0000: ┌ Link step
➤ YN0000: │ ESM support for PnP uses the experimental loader API and is therefore experimental
➤ YN0000: └ Completed
➤ YN0000: Done with warnings in 0s 299ms
Error: Could not resolve @prisma/client despite the installation that we just tried.
Please try to install it by hand with yarn add @prisma/client and rerun npx prisma generate 🙏.
```

**Reply by pantharshit00** [reply]
Hello @yuranos,

Can you please open a [new discussion](https://github.com/prisma/prisma/discussions/new) describing your environment? Can you also please share debug output in that: https://www.prisma.io/docs/guides/database/troubleshooting-orm/creating-bug-reports#include-logging-and-debugging-output?

**Comment by malindupasan** [comment]
Remove prisma directory,nodemodules,package.json from parent directory

**Reply by sparkidea25** [reply]
where do we both to?

**Comment by Marlrus** [comment]
I ran into this issue as well in a GitHub action but strangely not when running the exact same commands locally.

I had the exact same error message:

```
Prisma schema loaded from prisma/schema.prisma

up to date, audited 82 packages in 2s

6 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
Error: Could not resolve @prisma/client despite the installation that we just tried.
Please try to install it by hand with npm i @prisma/client and rerun npx "prisma generate" 🙏.
```

## Context and Environment

Running an `nx` monorepo with 3 apps in it. One of the apps uses Prisma.

The App that uses Prisma is using the `aws cdk` for deployments and uses the Prisma Client in `AWS Lambdas` that run in a `node` environment. The error happened during the building bundling step where the CDK bundles the TypeScript code and creates the `zip` file that will be shipped into the lambda.

## Sumary

I solved this by building the Prisma Client in the root directory (Single `node_modules` directory) and then copying the output into the `node_modules` of the lambdas being bundled.

## CDK Lambda Config

To build the handler I'm using the `NodejsFunction` construct from the AWS CDK.
This takes a `bundling` configuration where you exclude `nodeModules` from being bundled by `esbuild`. There are `commandHooks` that let you run operations before bundling, before installing, and after bundling.

### Issue

I used to have the `@prisma/client` package excluded through `nodeModules` which mean that the bundle directory had the `@prisma/client` in its `node_modules` directory.

In the Before Install step, I would copy the `prisma/schema.prisma` into the bundle directory.

In the After Bundling step, I would run `npx prisma generate` which would be the step where the Action Failed.

### Solution

The Pipeline runs `npm run prismagenerate` which is a `prisma generate --schema=<TARGET_DIR>` command.

I removed `@prisma/client` from exclusion, meaning that the lambda _will not have a `@prisma/client` in the `node_modules` directory_.

I'm not copying the `prisma/schema.prisma` file into the bundle directory.

In After Bundling I'm `rm -rf` anything `node_modules/@prisma` and `node_modules/.prisma` from the bundle directory. Then I copy `node_modules/@prisma` and `node_modules/.prisma` **from the root into the bundle directory**.

After the bundling is complete, I delete the `@prisma/engines` directory and the debian binary client from the `.prisma/clients` directory to reduce the bundle size being shipped in the `.zip` file to AWS Lambda.

**Comment by ciezczakmichal** [comment]
I encountered same error during transision from Yarn 1 to Yarn 4. I was building Nest app image using Docker.

In my case, the error was caused by the missing `.yarnrc.yml` file (containing `nodeLinker: node-modules`) in the container while installing dependencies.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
